### PR TITLE
ci(*:skip) Improve automated changelog generation

### PR DESCRIPTION
**Pull request by charlesbvll**
### Description

Our changelog generation script didn't work automatically with our new PR title format.

### Related issues/PRs

N/A

### Explanation

Use the PR title to categorise the PRs.

### Checklist

- [x] Implement proposed change
- [x] Make CI checks pass
- [x] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### Any other comments?

N/A

### CODE REVIEW

File: dev/update_changelog.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+from github import Github
+from github.PullRequest import PullRequest
+from github.Repository import Repository
+from github.Tag import Tag
+
+REPO_NAME = "adap/flower"
+CHANGELOG_FILE = "doc/source/ref-changelog.md"
+CHANGELOG_SECTION_HEADER = "### Changelog entry"
+
+# Load the YAML configuration
```
Comment (danieljanes) [inline dev/update_changelog.py:35]:
```suggestion
# Load the TOML configuration
```

**Review by danieljanes — COMMENTED** [review]
Instead of moving `update_changelog` to `dev`, it would be better to keep it in `flwr_tool` and eventually move `flwr_tool` into a separate project

**Comment by charlesbvll** [comment]
> Instead of moving `update_changelog` to `dev`, it would be better to keep it in `flwr_tool` and eventually move `flwr_tool` into a separate project

Shouldn't this be done in a separate PR? I feel like in terms of consistency, it makes more sense to have everything under `dev/` (as `flwr_tool` currently only contains the `protoc.py` script, whereas `dev/` contains all dev scripts similar to the `update_changelog.py` script)

**Comment by danieljanes** [comment]
The "move `flwr_tool` into a new project" should happen separately. But we had `update_changelog.py` in `flwr_tool` already (this PR removes it), so I'd recommend to leave it there and make the changes in the version in `flwr_tool`.

**Comment by charlesbvll** [comment]
> The "move `flwr_tool` into a new project" should happen separately. But we had `update_changelog.py` in `flwr_tool` already (this PR removes it), so I'd recommend to leave it there and make the changes in the version in `flwr_tool`.

@danieljanes But this new script uses the same `changelog_config.toml` file as does `check_pr_title` script. This is why it was moved to `dev/` so we don't have to do some relative path handling

File: doc/source/contributor-how-to-release-flower.rst [hunk 2]
```diff
@@ -9,7 +9,7 @@ During the release
 The version number of a release is stated in ``pyproject.toml``. To release a new version of Flower, the following things need to happen (in that order):
 
 1. Run ``python3 src/py/flwr_tool/update_changelog.py <YOUR_GH_TOKEN>`` in order to add every new change to the changelog (feel free to make manual changes to the changelog afterwards until it looks good).
```
Comment (danieljanes) [inline doc/source/contributor-how-to-release-flower.rst:11]:
```suggestion
1. Run ``python3 ./dev/update_changelog.py <YOUR_GH_TOKEN>`` in order to add every new change to the changelog (feel free to make manual changes to the changelog afterwards until it looks good).
```

File: dev/update_changelog.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Update the changelog using PR titles."""
+
+
+import pathlib
+import re
+import tomllib
```
Comment (panh99) [inline dev/update_changelog.py:21]: I think this won't work in Python <3.11. We should use `import tomli as tomllib` when the python version is less than 3.11.

File: dev/update_changelog.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    _update_changelog(prs)
+
+    new_version = _bump_minor_version(latest_tag)
+    if not new_version:
+        print("Wrong tag format.")
+        return
+
+    _add_shorlog(new_version, shortlog)
+
+    print("Changelog updated succesfully.")
```
Comment (panh99) [inline dev/update_changelog.py:286]: Maybe we should always print this? If there's no `## Unreleased` tag, we will see:

```
Unreleased header not found in the changelog.
Changelog updated succesfully.
```

File: dev/update_changelog.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+    gh_api = Github(argv[1])
+    repo, latest_tag = _get_latest_tag(gh_api)
+    if not latest_tag:
+        print("No tags found in the repository.")
+        return
+
+    shortlog, prs = _get_pull_requests_since_tag(repo, latest_tag)
+    _update_changelog(prs)
+
+    new_version = _bump_minor_version(latest_tag)
```
[the remark this answers is not in the document]
Comment (charlesbvll) [reply dev/update_changelog.py:284]: No, and this is intentional, this script is meant to be ran multiple times and to only modify the changelog

**Pull request merged**